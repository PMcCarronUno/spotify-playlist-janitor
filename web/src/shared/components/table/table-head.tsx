import { useState } from "react";
import styled from "styled-components";
import { VscArrowUp, VscArrowDown } from "react-icons/vsc";

import { TABLE_HOVER } from "shared/constants";

import { Column, SortOrder } from "./table-types";
import { Text } from "../typography";

type TableHeadProps = {
  columns: Column[];
  handleSorting(accessor: string, sortOrder: SortOrder): void;
};

const Tr = styled.tr`
  .sortable {
    cursor: pointer !important;
  }

  .sortable:hover {
    background-color: ${TABLE_HOVER};
  }

  .content {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
  }

  svg {
    padding-top: 3px;
    margin-bottom: -3px;
  }
`;

export const TableHead = ({ columns, handleSorting }: TableHeadProps) => {
  const [sortField, setSortField] = useState("");
  const [order, setOrder] = useState<SortOrder>("asc");

  const handleSortingChange = (accessor: any) => {
    const sortOrder =
      accessor === sortField && order === "asc" ? "desc" : "asc";
    setSortField(accessor);
    setOrder(sortOrder);
    handleSorting(accessor, sortOrder);
  };

  return (
    <thead>
      <Tr>
        {columns.map(({ label, accessor, sortable }: any) => {
          let icon = undefined;

          if (accessor === sortField) {
            icon = order === "asc" ? <VscArrowUp /> : <VscArrowDown />;
          }
          return (
            <th
              key={accessor}
              onClick={
                sortable ? () => handleSortingChange(accessor) : () => {}
              }
              className={sortable ? "sortable" : ""}
            >
              <div className="content">
                <Text>{label}</Text>
                {icon}
              </div>
            </th>
          );
        })}
      </Tr>
    </thead>
  );
};

TableHead.displayName = "TableHead";
