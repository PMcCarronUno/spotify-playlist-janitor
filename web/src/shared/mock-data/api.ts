import { LoginResponse } from "api/auth-api/auth-api-types";
import { DatabasePlaylistResponse, SkippedTrackResponse } from "api/data-api";
import { SpotifyPlaylistResponse } from "api/spotify-api";

type AuthData = {
  login: LoginResponse;
};

export const auth: AuthData = {
  login: {
    refresh_token: "access_token",
    token_type: "Bearer",
    access_token: "access_token",
    expires_in: 4503,
  },
};

export const databasePlaylists: DatabasePlaylistResponse[] = [
  {
    id: "1",
    skipThreshold: 10,
    ignoreInitialSkips: false,
    autoCleanupLimit: 10,
  },
  {
    id: "2",
    skipThreshold: 2,
    ignoreInitialSkips: false,
    autoCleanupLimit: 2,
  },
  {
    id: "3",
    skipThreshold: 3,
    ignoreInitialSkips: true,
    autoCleanupLimit: 3,
  },
  {
    id: "4",
    skipThreshold: 4,
    ignoreInitialSkips: true,
    autoCleanupLimit: 4,
  },
];

export const spotifyPlaylists: SpotifyPlaylistResponse[] = [
  {
    id: "1",
    name: "Playlist 1",
    href: "https://open.spotify.com/playlist/1",
    images: [
      {
        height: 640,
        width: 640,
        url: "https://mosaic.scdn.co/101",
      },
      {
        height: 300,
        width: 300,
        url: "https://mosaic.scdn.co/102",
      },
      {
        height: 60,
        width: 60,
        url: "https://mosaic.scdn.co/103",
      },
    ],
  },
  {
    id: "2",
    name: "Playlist 2",
    href: "https://open.spotify.com/playlist/2",
    images: [
      {
        height: 640,
        width: 640,
        url: "https://mosaic.scdn.co/201",
      },
      {
        height: 300,
        width: 300,
        url: "https://mosaic.scdn.co/202",
      },
      {
        height: 60,
        width: 60,
        url: "https://mosaic.scdn.co/203",
      },
    ],
  },
  {
    id: "3",
    name: "Playlist 3",
    href: "https://open.spotify.com/playlist/3",
    images: [
      {
        height: 640,
        width: 640,
        url: "https://mosaic.scdn.co/301",
      },
      {
        height: 300,
        width: 300,
        url: "https://mosaic.scdn.co/302",
      },
      {
        height: 60,
        width: 60,
        url: "https://mosaic.scdn.co/303",
      },
    ],
  },
  {
    id: "4",
    name: "Playlist 4",
    href: "https://open.spotify.com/playlist/4",
    images: [
      {
        height: 640,
        width: 640,
        url: "https://mosaic.scdn.co/401",
      },
      {
        height: 300,
        width: 300,
        url: "https://mosaic.scdn.co/402",
      },
      {
        height: 60,
        width: 60,
        url: "https://mosaic.scdn.co/403",
      },
    ],
  },
  {
    id: "5",
    name: "Playlist 5",
    href: "https://open.spotify.com/playlist/5",
    images: [
      {
        height: 640,
        width: 640,
        url: "https://mosaic.scdn.co/501",
      },
      {
        height: 300,
        width: 300,
        url: "https://mosaic.scdn.co/502",
      },
      {
        height: 60,
        width: 60,
        url: "https://mosaic.scdn.co/503",
      },
    ],
  },
  {
    id: "6",
    name: "Playlist 6",
    href: "https://open.spotify.com/playlist/6",
    images: [
      {
        height: 640,
        width: 640,
        url: "https://mosaic.scdn.co/601",
      },
      {
        height: 300,
        width: 300,
        url: "https://mosaic.scdn.co/602",
      },
      {
        height: 60,
        width: 60,
        url: "https://mosaic.scdn.co/603",
      },
    ],
  },
];

export const skippedTracks: SkippedTrackResponse = [
  {
    playlistId: databasePlaylists[0].id,
    trackId: "track 01",
    name: "Name 1",
    skippedDate: new Date("2023-10-31T19:51:25Z"),
    artists: [
      {
        id: "artist 01",
        name: "Artist 1",
        href: "https://open.spotify.com/artist/1",
      }
    ],
    album: {
      id: "album 01",
      name: "Album 1",
      href: "https://open.spotify.com/album/1",
      images: [
        {
          height: 240,
          width: 240,
          url: "https://mosaic.scdn.co/101",
        }
      ]
    }
  },
  {
    playlistId: databasePlaylists[0].id,
    trackId: "track 01",
    name: "Name 1",
    skippedDate: new Date("2023-10-31T19:51:38Z"),
    artists: [
      {
        id: "artist 01",
        name: "Artist 1",
        href: "https://open.spotify.com/artist/1",
      }
    ],
    album: {
      id: "album 01",
      name: "Album 1",
      href: "https://open.spotify.com/album/1",
      images: [
        {
          height: 240,
          width: 240,
          url: "https://mosaic.scdn.co/101",
        }
      ]
    }
  },
  {
    playlistId: databasePlaylists[0].id,
    trackId: "track 01",
    name: "Name 1",
    skippedDate: new Date("2023-10-31T19:52:07Z"),
    artists: [
      {
        id: "artist 01",
        name: "Artist 1",
        href: "https://open.spotify.com/artist/1",
      }
    ],
    album: {
      id: "album 01",
      name: "Album 1",
      href: "https://open.spotify.com/album/1",
      images: [
        {
          height: 240,
          width: 240,
          url: "https://mosaic.scdn.co/101",
        }
      ]
    }
  },
  {
    playlistId: databasePlaylists[0].id,
    trackId: "track 02",
    name: "Name 2",
    skippedDate: new Date("2023-10-31T19:53:10Z"),
    artists: [
      {
        id: "artist 02",
        name: "Artist 2",
        href: "https://open.spotify.com/artist/2",
      }
    ],
    album: {
      id: "album 01",
      name: "Album 2",
      href: "https://open.spotify.com/album/2",
      images: [
        {
          height: 240,
          width: 240,
          url: "https://mosaic.scdn.co/201",
        }
      ]
    }
  },
  {
    playlistId: databasePlaylists[0].id,
    trackId: "track 02",
    name: "Name 2",
    skippedDate: new Date("2023-10-31T19:53:30Z"),
    artists: [
      {
        id: "artist 02",
        name: "Artist 2",
        href: "https://open.spotify.com/artist/2",
      }
    ],
    album: {
      id: "album 01",
      name: "Album 2",
      href: "https://open.spotify.com/album/2",
      images: [
        {
          height: 240,
          width: 240,
          url: "https://mosaic.scdn.co/201",
        }
      ]
    }
  }
];
