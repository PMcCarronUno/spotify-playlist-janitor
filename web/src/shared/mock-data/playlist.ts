import { DetailedPlaylist } from "modules/playlist/playlist-types";
import {
  SkippedTrackHistory,
  SkippedTrackTotal,
  Track,
} from "modules/playlist/tabs/playlist-tabs-types";
import { DEFAULT_ALBUM_IMAGE } from "shared/constants";
import {
  databasePlaylists,
  spotifyPlaylists,
  skippedTracks,
  spotifyTracks as mockSpotifyTracks,
} from "shared/mock-data/api";
import { Image } from "shared/types";

const image: Image = {
  height:
    spotifyPlaylists.find((x) => x.id === databasePlaylists[0].id)?.images[0]
      ?.height || 100,
  width:
    spotifyPlaylists.find((x) => x.id === databasePlaylists[0].id)?.images[0]
      ?.width || 100,
  url:
    spotifyPlaylists.find((x) => x.id === databasePlaylists[0].id)?.images[0]
      ?.url || "href",
};

export const playlist: DetailedPlaylist = {
  id: databasePlaylists[0].id,
  href:
    spotifyPlaylists.find((x) => x.id === databasePlaylists[0].id)?.href || "",
  name:
    spotifyPlaylists.find((x) => x.id === databasePlaylists[0].id)?.name || "",
  image: image,
  ignoreInitialSkips: databasePlaylists[0].ignoreInitialSkips,
  autoCleanupLimit: databasePlaylists[0].autoCleanupLimit,
  skipThreshold: databasePlaylists[0].skipThreshold,
};

export const skippedTrackTotals: SkippedTrackTotal[] = [
  {
    id: skippedTracks[0].trackId,
    name: skippedTracks[0].name,
    duration: skippedTracks[0].duration,
    skippedTotal: 3,
    album: skippedTracks[0].album,
    artists: skippedTracks[0].artists,
    image: skippedTracks[0].album.images.reduce((prev, curr) => {
      return prev.height < curr.height ? prev : curr;
    }),
  },
  {
    id: skippedTracks[3].trackId,
    name: skippedTracks[3].name,
    duration: skippedTracks[3].duration,
    skippedTotal: 2,
    album: skippedTracks[3].album,
    artists: skippedTracks[3].artists,
    image: skippedTracks[3].album.images.reduce((prev, curr) => {
      return prev.height < curr.height ? prev : curr;
    }),
  },
];

export const skippedTrackHistory: SkippedTrackHistory[] = skippedTracks.map(
  (skippedTrack) => ({
    id: skippedTrack.trackId,
    name: skippedTrack.name,
    duration: skippedTrack.duration,
    skippedDate: new Date(skippedTrack.skippedDate),
    album: skippedTrack.album,
    artists: skippedTrack.artists,
    image: skippedTrack.album.images.reduce((prev, curr) => {
      return prev.height < curr.height ? prev : curr;
    }),
  })
);

export const spotifyTracks: Track[] = mockSpotifyTracks.map((spotifytrack) => {
  let image: Image = {
    height: 40,
    width: 40,
    url: DEFAULT_ALBUM_IMAGE,
  };

  if (spotifytrack.album.images.length > 0) {
    image = spotifytrack.album.images.reduce((prev, curr) => {
      return prev.height < curr.height ? prev : curr;
    });
  }

  return {
    id: spotifytrack.id,
    name: spotifytrack.name,
    duration: spotifytrack.duration,
    album: spotifytrack.album,
    artists: spotifytrack.artists,
    image,
  };
});
