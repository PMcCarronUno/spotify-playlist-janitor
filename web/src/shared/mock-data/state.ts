import { UserToken } from "shared/types";

export const userState: UserToken = {
  refresh_token: "access_token",
  token_type: "Bearer",
  access_token: "access_token",
  expires_in: 4503,
  expires_on: new Date().getTime() + 1000 * 100,
};
