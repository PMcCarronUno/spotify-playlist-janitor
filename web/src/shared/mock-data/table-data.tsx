import { Column } from "../components/table";
import { PiClock } from "react-icons/pi";
import { VscTrash } from "react-icons/vsc";

export const tableColumns: Column[] = [
  {
    label: "",
    accessor: "image",
    sortable: false,
    primary: true,
    render: (url: string) => <img alt="" width={40} height={40} src={url} />,
  },
  {
    label: "Title",
    accessor: "title",
    sortable: true,
    sortbyOrder: "desc",
    primary: true,
  },
  { label: "Artist", accessor: "artist", sortable: true },
  { label: "Album", accessor: "album", sortable: true },
  {
    label: <PiClock />,
    accessor: "length",
    sortable: true,
    render: (millis: number) => {
      var minutes = Math.floor(millis / 60000);
      var seconds = parseInt(((millis % 60000) / 1000).toFixed(0));
      return seconds === 60
        ? minutes + 1 + ":00"
        : minutes + ":" + (seconds < 10 ? "0" : "") + seconds;
    },
  },
  {
    label: "Date Skipped",
    accessor: "date_skipped",
    sortable: true,
    render: (date: Date) => date.toISOString(),
  },
  {
    label: "",
    accessor: "delete",
    sortable: false,
    render: (id: string) => (
      <VscTrash
        style={{ cursor: "pointer" }}
        onClick={() => alert(`Deleted track id: ${id}`)}
      />
    ),
  },
];

export const tableData = [
  {
    id: 1,
    image: "https://i.scdn.co/image/ab67616d00004851835a9e77dae1c928f871ac73",
    title: "Longshot",
    artist: "Catfish and the Bottlemen",
    album: "The Balance",
    length: 232960,
    date_skipped: new Date(2023, 9, 15, 15, 30, 45, 20),
    delete: undefined,
  },
  {
    id: 2,
    image: "https://i.scdn.co/image/ab67616d0000485158406b3f1ac3ceaff7a64fef",
    title: "Dark Necessities",
    artist: "Red Hot Chili Peppers",
    album: "The Getaway",
    length: 302000,
    date_skipped: new Date(2023, 9, 15, 15, 30, 46, 25),
    delete: undefined,
  },
  {
    id: 3,
    image: "https://i.scdn.co/image/ab67616d00004851d86a1e021e7acc7c07c7d668",
    title: "Live Forever",
    artist: "Oasis",
    album: "Definitely Maybe",
    length: 302000,
    date_skipped: new Date(2023, 9, 15, 15, 30, 47, 30),
    delete: undefined,
  },
  {
    id: 4,
    image: "https://i.scdn.co/image/ab67616d00004851176ca6e2a6efb0e3aaf4e37a",
    title: "Touch It / Technologic",
    artist: "Daft Punk",
    album: "Alive 2007",
    length: 276666,
    date_skipped: new Date(2023, 9, 15, 15, 30, 47, 40),
    delete: undefined,
  },
];
