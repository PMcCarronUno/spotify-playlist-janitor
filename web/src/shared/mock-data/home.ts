import { Playlist } from "modules/home/home-types";
import { databasePlaylists, spotifyPlaylists } from "shared/mock-data/api";

export const monitoredPlaylists: Playlist[] = spotifyPlaylists
  .filter((spotifyPlaylist) =>
    databasePlaylists.some(
      (databasePlaylist) => databasePlaylist.id === spotifyPlaylist.id
    )
  )
  .map((monitoredPlaylist) => ({
    id: monitoredPlaylist.id,
    href: monitoredPlaylist.href,
    name: monitoredPlaylist.name,
    image: monitoredPlaylist.images[0],
  }));

export const unmonitoredPlaylists: Playlist[] = spotifyPlaylists
  .filter(
    (spotifyPlaylist) =>
      !databasePlaylists.some(
        (databasePlaylist) => databasePlaylist.id === spotifyPlaylist.id
      )
  )
  .map((monitoredPlaylist) => ({
    id: monitoredPlaylist.id,
    href: monitoredPlaylist.href,
    name: monitoredPlaylist.name,
    image: monitoredPlaylist.images[0],
  }));
