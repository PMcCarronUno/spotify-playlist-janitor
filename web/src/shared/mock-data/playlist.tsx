import format from "date-fns/format";
import { DetailedPlaylist } from "modules/playlist/playlist-types";
import {
  SkippedTrackAlbum,
  SkippedTrackArtist,
  SkippedTrackHistory,
  SkippedTrackTotal,
  Track,
} from "modules/playlist/tabs/playlist-tabs-types";
import { PiClock } from "react-icons/pi";
import { VscTrash } from "react-icons/vsc";
import { Skeleton, SkeletonTheme } from "shared/components/skeleton";
import { Column } from "shared/components/table";
import { SecondaryText } from "shared/components/typography";
import {
  DEFAULT_ALBUM_IMAGE,
  SKELETON_HIGHLIGHT,
  SKELETON_PLAYLIST_BASE,
} from "shared/constants";
import {
  databasePlaylists,
  spotifyPlaylists,
  skippedTracks,
  spotifyTracks as mockSpotifyTracks,
} from "shared/mock-data/api";
import { Image } from "shared/types";

const image: Image = {
  height:
    spotifyPlaylists.find((x) => x.id === databasePlaylists[0].id)?.images[0]
      ?.height || 100,
  width:
    spotifyPlaylists.find((x) => x.id === databasePlaylists[0].id)?.images[0]
      ?.width || 100,
  url:
    spotifyPlaylists.find((x) => x.id === databasePlaylists[0].id)?.images[0]
      ?.url || "href",
};

export const playlist: DetailedPlaylist = {
  id: databasePlaylists[0].id,
  href:
    spotifyPlaylists.find((x) => x.id === databasePlaylists[0].id)?.href || "",
  name:
    spotifyPlaylists.find((x) => x.id === databasePlaylists[0].id)?.name || "",
  image: image,
  ignoreInitialSkips: databasePlaylists[0].ignoreInitialSkips,
  autoCleanupLimit: databasePlaylists[0].autoCleanupLimit,
  skipThreshold: databasePlaylists[0].skipThreshold,
};

export const skippedTrackTotals: SkippedTrackTotal[] = [
  {
    id: skippedTracks[0].trackId,
    name: skippedTracks[0].name,
    duration: skippedTracks[0].duration,
    skippedTotal: 3,
    album: skippedTracks[0].album,
    artists: skippedTracks[0].artists,
    image: skippedTracks[0].album.images.reduce((prev, curr) => {
      return prev.height < curr.height ? prev : curr;
    }),
  },
  {
    id: skippedTracks[3].trackId,
    name: skippedTracks[3].name,
    duration: skippedTracks[3].duration,
    skippedTotal: 2,
    album: skippedTracks[3].album,
    artists: skippedTracks[3].artists,
    image: skippedTracks[3].album.images.reduce((prev, curr) => {
      return prev.height < curr.height ? prev : curr;
    }),
  },
];

export const skippedTrackHistory: SkippedTrackHistory[] = skippedTracks.map(
  (skippedTrack) => ({
    id: skippedTrack.trackId,
    name: skippedTrack.name,
    duration: skippedTrack.duration,
    skippedDate: new Date(skippedTrack.skippedDate),
    album: skippedTrack.album,
    artists: skippedTrack.artists,
    image: skippedTrack.album.images.reduce((prev, curr) => {
      return prev.height < curr.height ? prev : curr;
    }),
  })
);

export const spotifyTracks: Track[] = mockSpotifyTracks.map((spotifytrack) => {
  let image: Image = {
    height: 40,
    width: 40,
    url: DEFAULT_ALBUM_IMAGE,
  };

  if (spotifytrack.album.images.length > 0) {
    image = spotifytrack.album.images.reduce((prev, curr) => {
      return prev.height < curr.height ? prev : curr;
    });
  }

  return {
    id: spotifytrack.id,
    name: spotifytrack.name,
    duration: spotifytrack.duration,
    album: spotifytrack.album,
    artists: spotifytrack.artists,
    image,
  };
});

const skeleton = (
  <SkeletonTheme
    baseColor={SKELETON_PLAYLIST_BASE}
    highlightColor={SKELETON_HIGHLIGHT}
    height="100%"
  >
    <Skeleton />
  </SkeletonTheme>
);

export const imageColumn: Column = {
  label: "",
  accessor: "image",
  primary: true,
  render: (image: Image, loading?: boolean) => (
    <div className="centered">
      {loading ? (
        <div style={{ height: 40, width: 40 }}>{skeleton}</div>
      ) : (
        <img alt="" width={40} height={40} src={image.url} />
      )}
    </div>
  ),
};

export const titleColumn: Column = {
  label: "Title",
  accessor: "name",
  sortable: true,
  primary: true,
  render: (value: string, loading?: boolean) => (loading ? skeleton : value),
};

export const artistColumn: Column = {
  label: "Artist",
  accessor: "artists",
  sortable: true,
  render: (artists: SkippedTrackArtist[], loading?: boolean) => (
    <>
      {loading
        ? skeleton
        : artists.map((artist, index) => (
            <>
              {`${index === 0 ? "" : ", "}`}
              {!!artist.href ? (
                <a href={artist.href} target="_blank" rel="noreferrer">
                  {artist.name}
                </a>
              ) : (
                <SecondaryText>{artist.name}</SecondaryText>
              )}
            </>
          ))}
    </>
  ),
};

export const albumColumn: Column = {
  label: "Album",
  accessor: "album",
  sortable: true,
  render: (album: SkippedTrackAlbum, loading?: boolean) =>
    loading ? (
      skeleton
    ) : !!album.href ? (
      <a href={album.href} target="_blank" rel="noreferrer">
        {album.name}
      </a>
    ) : (
      <SecondaryText>{album.name}</SecondaryText>
    ),
};

export const durationColumn: Column = {
  label: <PiClock />,
  accessor: "duration",
  sortable: true,
  render: (millis: number, loading?: boolean) => {
    if (loading) {
      return skeleton;
    }
    var minutes = Math.floor(millis / 60000);
    var seconds = parseInt(((millis % 60000) / 1000).toFixed(0));
    return seconds === 60
      ? minutes + 1 + ":00"
      : minutes + ":" + (seconds < 10 ? "0" : "") + seconds;
  },
};

export const totalSkippedColumn: Column = {
  label: "Date Skipped",
  accessor: "skippedDate",
  sortable: true,
  sortbyOrder: "desc",
  render: (date: Date, loading?: boolean) =>
    loading
      ? skeleton
      : format(date, "MMM dd yyyy") + " at " + format(date, "HH:mm:ss"),
};

export const dateSkippedColumn: Column = {
  label: "Total Skipped",
  accessor: "skippedTotal",
  sortable: true,
  sortbyOrder: "desc",
  render: (value: string, loading?: boolean) => (loading ? skeleton : value),
};

export const getDeleteColumn = (onDeleteClick: (id: string) => void) => ({
  label: "",
  accessor: "id",
  render: (id: string, loading?: boolean) =>
    loading ? (
      <></>
    ) : (
      <div className="centered">
        <VscTrash
          style={{ cursor: "pointer" }}
          onClick={() => onDeleteClick(id)}
        />
      </div>
    ),
});
