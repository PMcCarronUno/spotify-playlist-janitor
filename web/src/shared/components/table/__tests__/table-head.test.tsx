import {
  fireEvent,
  getByTestId,
  getByText,
  render,
} from "@testing-library/react";
import { tableColumns } from "shared/mock-data/table-data";

import { TableHead } from "../table-head";

describe("<TableHead />", () => {
  let container: HTMLElement;
  let props: any;
  let mockHandleSort = jest.fn();

  beforeEach(() => {
    props = {
      columns: tableColumns,
      handleSorting: mockHandleSort,
    };

    ({ container } = render(<TableHead {...props} />));
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("should render TableHead component", () => {
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should execute handleSort when sortable header clicked", () => {
    const colunHead = getByTestId(container, "table-column-head-title");
    fireEvent.click(colunHead);

    expect(mockHandleSort).toHaveBeenCalledWith("title", "asc");
  });

  it("should execute handleSort when sortable header clicked with opposite sort order", () => {
    const colunHead = getByTestId(container, "table-column-head-title");
    fireEvent.click(colunHead);
    fireEvent.click(colunHead);

    expect(mockHandleSort).toHaveBeenNthCalledWith(1, "title", "asc");
    expect(mockHandleSort).toHaveBeenLastCalledWith("title", "desc");
  });

  it("should not execute handleSort when non-sortable header clicked", () => {
    const colunHead = getByTestId(container, "table-column-head-image");
    fireEvent.click(colunHead);

    expect(mockHandleSort).not.toHaveBeenCalled();
  });
});
