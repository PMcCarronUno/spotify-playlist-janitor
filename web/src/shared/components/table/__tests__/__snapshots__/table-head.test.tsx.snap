// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TableHead /> should render TableHead component 1`] = `
<table>
  <thead>
    <tr
      class="sc-jTQDJr hyFECk"
    >
      <th
        class=""
        data-testid="table-head-cell-image"
        id="table-head-cell-image"
      >
        <div
          class="content"
        >
          <div
            class="sc-dmyDmy dcXriK"
          />
        </div>
      </th>
      <th
        class="sortable"
        data-testid="table-head-cell-title"
        id="table-head-cell-title"
      >
        <div
          class="content"
        >
          <div
            class="sc-dmyDmy dcXriK"
          >
            Title
          </div>
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M3.147 9l5 5h.707l5-5-.707-.707L9 12.439V2H8v10.44L3.854 8.292 3.147 9z"
              fill-rule="evenodd"
            />
          </svg>
        </div>
      </th>
      <th
        class="sortable"
        data-testid="table-head-cell-artist"
        id="table-head-cell-artist"
      >
        <div
          class="content"
        >
          <div
            class="sc-dmyDmy dcXriK"
          >
            Artist
          </div>
        </div>
      </th>
      <th
        class="sortable"
        data-testid="table-head-cell-album"
        id="table-head-cell-album"
      >
        <div
          class="content"
        >
          <div
            class="sc-dmyDmy dcXriK"
          >
            Album
          </div>
        </div>
      </th>
      <th
        class="sortable"
        data-testid="table-head-cell-length"
        id="table-head-cell-length"
      >
        <div
          class="content"
        >
          <div
            class="sc-dmyDmy dcXriK"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 256 256"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M128,24A104,104,0,1,0,232,128,104.11,104.11,0,0,0,128,24Zm0,192a88,88,0,1,1,88-88A88.1,88.1,0,0,1,128,216Zm64-88a8,8,0,0,1-8,8H128a8,8,0,0,1-8-8V72a8,8,0,0,1,16,0v48h48A8,8,0,0,1,192,128Z"
              />
            </svg>
          </div>
        </div>
      </th>
      <th
        class="sortable"
        data-testid="table-head-cell-date_skipped"
        id="table-head-cell-date_skipped"
      >
        <div
          class="content"
        >
          <div
            class="sc-dmyDmy dcXriK"
          >
            Date Skipped
          </div>
        </div>
      </th>
      <th
        class=""
        data-testid="table-head-cell-delete"
        id="table-head-cell-delete"
      >
        <div
          class="content"
        >
          <div
            class="sc-dmyDmy dcXriK"
          />
        </div>
      </th>
    </tr>
  </thead>
</table>
`;
